src/advanced/timestamping-tutorial.md

```diff
@@ -0,0 +1,929 @@
+    fn service_id(&self) -> u16 {
+        TIMESTAMPING_SERVICE
+    }
+
+    fn service_name(&self) -> &'static str {
+        SERVICE_NAME
+    }
+
+    fn state_hash(&self, view: &dyn Snapshot) -> Vec<Hash> {
+        let schema = Schema::new(view);
+        schema.state_hash()
+    }
+
+    fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<dyn Transaction>, StreamStructError> {
+        let tx = TimeTransactions::tx_from_raw(raw)?;
+        Ok(tx.into())
+    }
+
+    fn wire_api(&self, builder: &mut ServiceApiBuilder) {
+        PublicApi::wire(builder);
+    }
+}
+```
+The code above declares a number of methods:
+- `service_id` - returns the ID of the service
+- `service_name` - returns the name of the service
+- `state_hash` - calculates the hash of
+  [the blockchain state](../glossary.md#blockchain-state). The method
+  should return [a vector of hashes](../architecture/services.md#state-hash)
+  of the [Merkelized service tables](../glossary.md#merkelized-indices).
+- `tx_from_raw` - deserializes transactions processed by the service.
+   If the incoming transaction is built successfully, we put it into a `Box<_>`.
+- `wire_api` - connects the API endpoints of the service to the API of the
+  blockchain.
+
+## Create Service Instance
+
+We define the `ServiceFactory` structure in [src/lib.rs][src/lib.rs]. This
+structure builds an instance of our service.
+
+```rust
+#[derive(Debug, Clone, Copy)]
+pub struct ServiceFactory;
+```
+The `ServiceFactory` structure compiles an instance of the Timestamping service
+from the code we have written previously.
+
+```rust
+impl fabric::ServiceFactory for ServiceFactory {
+    fn service_name(&self) -> &str {
+        SERVICE_NAME
+    }
+
+    fn make_service(&mut self, _: &fabric::Context) -> Box<dyn blockchain::Service> {
+        Box::new(Service::new())
+    }
+}
+```
+The `ServiceFactory` returns the name of the service and includes the
+`make_service` function which will later allow us to include the new service
+into the `NodeBuilder`.
+
+## Create Builder for Exonum Node With Timestamping
+
+As the final step of configuring the backend of our Timestamping service, we
+add the [src/main.rs][src/main.rs] file, which lets us launch an Exonum node
+with additional services.
+
+In this file we import Exonum, its Configuration service, the Time oracle which
+will provide the currect time values and the Timestamping service itself.
+
+```rust
+fn main() {
+    exonum::helpers::init_logger().unwrap();
+    NodeBuilder::new()
+        .with_service(Box::new(exonum_configuration::ServiceFactory))
+        .with_service(Box::new(exonum_time::TimeServiceFactory))
+        .with_service(Box::new(exonum_timestamping::ServiceFactory))
+        .run();
+}
+```
+
+When called, the `main` function will lauch an Exonum node with the
+Configuration, Time and Timestamping services.
+
+!!! note
+    The services should be defined in exactly the same order as in the `main`
```

Comment:
I am soooorry, but who's come up with this idea? It seems completely wrong to me. The order of `with_service` methods invocations must not have any impact on actual start sequence of the services.